### PR TITLE
ref: geoip match

### LINKED ISSUE #161: Fix geoip logic
The logic should be as follows.

City:
1. Find the best city match based on the number of the same values (this is implemented correctly).
1.1. If ipinfo + maxmind have the same result, use ipinfo (this matters because we then use other values like coordinates/asn from the "best match" and they can differ even between providers with the same city).
1.2. If ipinfo + fastly have the same result, use ipinfo. 
1.3. If maxmind + fastly have the same result, use maxmind,
1.4. If all three providers return a different city, the priority is: ipinfo, maxmind (right now it's random).
1.5. If Fastly is the only to return a city, consider the location unresolved and drop the probe.

ASN/Network:
1. If ipinfo and maxmind agreed on the city, use ipinfo.
1.1. If ipinfo doesn't have network data, use maxmind.
2. Use the provider with "best match" for the city.
2.1 If the best match for the city doesn't have network data, drop the probe (with the exception of the specific case 1.1).

---

3. Handle a case when ipinfo doesn't return org, which currently throws: https://github.com/jsdelivr/globalping/blob/7b96ead590866a21613a35a3d2388dd7bf0c55ab/src/lib/geoip/providers/ipinfo.ts#L24
4. Test all the cases.
5. Increase default cache time to 3 days.

### PR DESCRIPTION
**Pull request by patrykcieszkowski**
resolve #161

### CODE REVIEW

File: src/lib/geoip/client.ts [hunk 1]
```diff
@@ -88,8 +94,9 @@ export default class GeoipClient {
 	}
 
 	private bestMatch(field: keyof LocationInfo, sources: LocationInfoWithProvider[]): LocationInfo {
-		const ranked = Object.values(_.groupBy(sources.filter(s => s[field]), field)).sort((a, b) => b.length - a.length).flat();
-		const best = ranked[0];
+		const grouped = Object.values(_.groupBy(sources.filter(s => s[field]), field));
+		const ranked = grouped.sort((a, b) => b.length - a.length).flat();
+		const best = grouped.length === sources.length ? sources.find(s => s.provider === 'ipinfo') : ranked[0];
```
[the remark this answers is not in the document]
Comment (patrykcieszkowski) [reply src/lib/geoip/client.ts:99]: filtering empty values before grouping them should do the trick

[the remark this answers is not in the document]
Comment (MartinKolarik) [reply src/lib/geoip/client.ts:99]: Yes but here you compare it to the `sources` which isn't filtered.

[the remark this answers is not in the document]
Comment (patrykcieszkowski) [reply src/lib/geoip/client.ts]: That's a good catch. I missed that. The latest fix should cover that.

File: src/lib/geoip/client.ts [hunk 2]
```diff
@@ -87,11 +101,38 @@ export default class GeoipClient {
...
+			};
+		}
+
+		const maxmind = sources.find(s => s.provider === 'maxmind');
+		if (maxmind?.asn && maxmind?.network) {
+			return {
+				asn: maxmind.asn,
+				network: maxmind.network,
+			};
+		}
```
Comment (MartinKolarik) [inline src/lib/geoip/client.ts:118]: Only if city for maxmind is the same as best match. If cities are different, we don't want to mix them and return undefined.

Comment (patrykcieszkowski) [reply src/lib/geoip/client.ts]: done

**Comment by MartinKolarik** [comment]
I think this point is still missing?
> Handle a case when ipinfo doesn't return org, which currently throws

**Comment by patrykcieszkowski** [comment]
What does it throw? In order for it to throw an error, it would have to be of type `undefined`, and I'm not seeing any reason why would it do that. It most likelly will just return an empty string.

**Pull request merged**